STY: fixes Pylint `C0209` Errors (uses f-strings)

HenryAsa opened the pull request:
I went through the `numpy` repository and updated all of the strings using `'{}'.format()` to be equivalent f-strings, resolving the [Pylint `C0209`](https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/consider-using-f-string.html) linting errors that were prevalent throughout the repository.  In terms of the functionality of the code, nothing was changed, but the code has been updated to meet more current Python style standards ([see Pylint `C0209` for more details](https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/consider-using-f-string.html)).

Resolves #26278

HenryAsa committed:
Addresses Pylint `C0209` linting errors

HenryAsa committed:
Fixed Pylint Error `C0209`

HenryAsa committed:
Updated minor linting errors

HenryAsa committed:
Fixed linting error `E225`

HenryAsa committed:
Reverted string back to original format

HenryAsa committed:
Minor bug fix with f-strings

charris commented:
I don't have a problem with this update, f strings are easier to read because they are more self documenting.

ngoldbaum commented on numpy/_core/arrayprint.py:
```diff
@@ -1432,9 +1432,9 @@ def __call__(self, x):
         ]
         if len(str_fields) == 1:
-            return "({},)".format(str_fields[0])
+            return f"({str_fields[0]},)".format(str_fields[0])
```
you left the `.format` behind

HenryAsa replied:
Good catch.  Thank you.  I fixed this.

ngoldbaum commented on numpy/_core/shape_base.py:
```diff
@@ -605,9 +602,9 @@ def _concatenate_shapes(shapes, axis):
+            f'Mismatched array shapes in block along axis {axis}.')
 
-    shape = (first_shape_pre + (sum(shape_at_axis),) + first_shape[axis+1:])
+    shape = first_shape_pre + (sum(shape_at_axis),) + first_shape[axis+1:]
```
this is an unrelated change but it looks correct to me, the parens aren't necessary

ngoldbaum commented on numpy/distutils/system_info.py:
```diff
@@ -2142,12 +2142,14 @@ def _calc_info_from_envvar(self):
         if len(unknown_order) > 0:
-            raise ValueError("blas_opt_info user defined BLAS order has unacceptable values: {}".format(unknown_order))
+            raise ValueError(
+                f"blas_opt_info user defined BLAS order "
```
this one doesn't need to be an f-string

HenryAsa replied:
Good catch.  Resolved.

ngoldbaum commented on numpy/f2py/crackfortran.py:
```diff
@@ -1480,7 +1480,7 @@ def analyzeline(m, case, line):
                                 expanded_list.append(match.strip())
                         matches = expanded_list
-                    new_val = "(/{}/)".format(", ".join(matches)) if vdim else matches[idx]
+                    new_val = f'(/{", ".join(matches)}/)' if vdim else matches[idx]
```
This and elsewhere in this file would be easier to read in this style if you bind `", ".join(matches)` to a variable name and use that here, I had to squint for a while trying to parse the nested quotes.

HenryAsa replied:
Good suggestion, would be happy to make those changes.  Since it is sometimes unclear what the code is doing at a certain section of the program, would it be alright if I use a "catch-all" variable name (perhaps `joined_elements`) to store these values?

In the cases where it is more clear what's being joined, I can name the variable something more specific to the operation that's occurring (like here, I would name the variable something along the lines of `all_matches` or `joined_matches`).  Or would you prefer that I use the same variable name (`joined_elements`) for all of these kinds of operations?

ngoldbaum replied:
Not a big deal, whatever you think is most understandable.

HenryAsa replied:
Okay, will do.

HenryAsa committed:
Fixed some f-strings to comply with Pylint `C0209`

HenryAsa committed:
Forgot to remove the `.format()` component of the string

HenryAsa committed:
Reverting the Makefile back to its original state

HenryAsa committed:
Updating f-strings to resolve Pylint `C0209`

HenryAsa committed:
Added more f-strings to resolve Pylint `C0209` errors

HenryAsa committed:
Resolved more Pylint `C0209` linting errors by converting strings to f-strings

HenryAsa committed:
Resolving more Pylint `C0209` errors

HenryAsa committed:
Resolved many Pylint `C0209` f-string errors

HenryAsa committed:
Fixed Pylint `C0209` errors by replacing strings with f-strings

HenryAsa committed:
Merge branch 'numpy:main' into main

ngoldbaum commented on numpy/_core/_internal.py:
```diff
@@ -905,11 +905,9 @@ def _ufunc_doc_signature_formatter(ufunc):
-        )
+        positional = ', '.join(
+                f'out{i+1}' for i in range(ufunc.nout))
+        out_args = f'[, {positional}], / [, out={repr((None,)*ufunc.nout)}]'
```
nit: binding a `default` variable as well would be more readable IMO

HenryAsa replied:
Yeah, good point.  I'll look into making it more readable.

ngoldbaum commented on numpy/_core/fromnumeric.py:
```diff
@@ -539,7 +539,7 @@ def put(a, ind, v, mode='raise'):
     except AttributeError as e:
         raise TypeError("argument 1 must be numpy.ndarray, "
-                        "not {name}".format(name=type(a).__name__)) from e
+                       f"not {type(a).__name__}") from e
```
nit: i'm pretty sure this is how e.g. black would align it

```suggestion
                        f"not {type(a).__name__}") from e
```

ngoldbaum reviewed (commented):
I started reading this but only got partway through before I ran out of time for this session of code review. I think the original version of this PR that I reviewed before was a lot shorter? It would be easier to review if you broke this up into multiple PRs that all converted a single "pattern" of usage.

Stepping back, the reason people tend to avoid refactors like this is because of the tendency to accidentally introduce bugs in code that might have used outdated idioms but was still working perfectly fine. It's also tedious to review code changes like this, especially if there's so many changes at once. There's a tendency to gloss over and not notice issues.

I would also suggest taking a look at automatic code upgraders like `ruff` or `pyupgrade` to do this work (I think you did this manually, please correct me if that's wrong). It would make it easier to review if you could just give me a `ruff` incantation to apply to get the exact PR diff you're trying to upstream, that way I don't have to worry about a human making a mistake somewhere.

ngoldbaum commented:
Oh also if we're going to make a global code change like this, we should probably add a linter rule to check for this.

HenryAsa commented:
> I started reading this but only got partway through before I ran out of time for this session of code review. I think the original version of this PR that I reviewed before was a lot shorter? It would be easier to review if you broke this up into multiple PRs that all converted a single "pattern" of usage.

Thanks for taking a look @ngoldbaum.  That makes sense.  Yes, the original pull request was slightly shorter; I adjusted the Regex expression I was using to find `.format()` strings to include more potential occurrences of these strings, and added those on to the pull request.  Funnily enough, that seems like a perfect "partition" I can use to break up these changes into multiple different pull-requests; I could partition the adjustments based on the Regex expression I used to find them.  Would that be preferable?
 
> Stepping back, the reason people tend to avoid refactors like this is because of the tendency to accidentally introduce bugs in code that might have used outdated idioms but was still working perfectly fine. It's also tedious to review code changes like this, especially if there's so many changes at once. There's a tendency to gloss over and not notice issues.

That makes sense.  I thought it could be convenient/cleaner to make these adjustments, but totally understand that they don't actually impact the day-to-day functionality of `numpy`.

> I would also suggest taking a look at automatic code upgraders like `ruff` or `pyupgrade` to do this work (I think you did this manually, please correct me if that's wrong). It would make it easier to review if you could just give me a `ruff` incantation to apply to get the exact PR diff you're trying to upstream, that way I don't have to worry about a human making a mistake somewhere.

Yes, all of these changes were done manually.  I took a lot of care to make sure I didn't accidentally change the resulting string, but I can see how this would be difficult to code review and catch potential bugs.  I'll look into those automated tools you mentioned.

rgommers commented:
> Yes, all of these changes were done manually. I took a lot of care to make sure I didn't accidentally change the resulting string, but I can see how this would be difficult to code review and catch potential bugs. I'll look into those automated tools you mentioned.

I suggest to close this then. Large diff for style-only changes are not desirable - a lot of work to review and the risk of regressions. A new PR with some selected tool-generated changes would be better. That said, please don't overdo it - style-only PRs are in general not the best thing to work on.

mattip commented:
Closing as per review.
